core/modules/replica/JobController.cc

```diff
@@ -0,0 +1,503 @@
+            maxReplicas,
+            computeCheckSum,
+            options
+        );
+
+    // Register the job (along with its callback) by its unique
+    // identifier in the local registry. Once it's complete it'll
+    // be automatically removed from the Registry.
+
+    _registry[job->id()] =
+        std::make_shared<JobWrapperImpl<VerifyJob>>(job, onFinish);
+
+    // Initiate the job
+    //
+    // FIXME: don't start the job right away. Put the request into the priority queue
+    // and call the scheduler's method to evaluate jobs in the queue to
+    // to see which should be started next (if any).
+
+    job->start();
+
+    return job;
+}
+
+DeleteWorkerJob::Ptr JobController::deleteWorker(
+                                            std::string const& worker,
+                                            bool permanentDelete,
+                                            DeleteWorkerJob::CallbackType onFinish,
+                                            Job::Options const& options) {
+
+    LOGS(_log, LOG_LVL_DEBUG, "JobController::deleteWorker");
+
+    util::Lock lock(_mtx, "JobController::deleteWorker");
+
+    auto const self = shared_from_this();
+    auto const job =
+        DeleteWorkerJob::create(
+            worker,
+            permanentDelete,
+            _controller,
+            std::string(),
+            [self] (DeleteWorkerJob::Ptr job) {
+                self->onFinish (job);
+            },
+            options
+        );
+
+    // Register the job (along with its callback) by its unique
+    // identifier in the local registry. Once it's complete it'll
+    // be automatically removed from the Registry.
+
+    _registry[job->id()] =
+        std::make_shared<JobWrapperImpl<DeleteWorkerJob>>(job, onFinish);
+
+    // Initiate the job
+    //
+    // FIXME: don't start the job right away. Put the request into the priority queue
+    // and call the scheduler's method to evaluate jobs in the queue to
+    // to see which should be started next (if any).
+
+    job->start();
+
+    return job;
+}
+
+void JobController::runQueued() {
+
+    LOGS(_log, LOG_LVL_DEBUG, "JobController::runQueued");
+
+    util::Lock lock(_mtx, "JobController::runQueued");
+
+    if (not isRunning()) return;
+
+    // TODO:
+    // Go through the input queue and evaluate which jobs should star
+    // now based on their scheduling criteria and on the status of
+    // the in-progres jobs (if any).
+    ;
+}
+
+void JobController::runScheduled() {
+
+    LOGS(_log, LOG_LVL_DEBUG, "JobController::runScheduled");
+
+    // Load the scheduled jobs (if any) from the database to see which ones
+    // need to be injected into the input queue.
+    //
+    // NOTE: don't prolifirate the lock's scope to avoid an imminent deadlock
+    //       when calling mehods which are called later.
+    {
+        util::Lock lock(_mtx, "JobController::runScheduled");
+
+        if (not isRunning()) return;
+
+        // TODO:
```

Comment:
todo what?
then it called run queued, but that function is a no-op. Seems like this needs some clean up and/or protection?

Reply:
The JobController has been redesigned. It's much simpler now. That method is gone.